torch/_functorch/aot_autograd.py

```diff
@@ -2036,6 +2095,298 @@ def wrapped_compiled_fn(args):
+        # to be concated into a larger primals list, instead of getting a list of lists
+        assert len(args) == 2
+        assert isinstance(args[0], list)
+        assert isinstance(args[1], list)
+        args_unwrapped_primals, wrap_info_primals = to_flat(args[0])
+        args_unwrapped_tangents, wrap_info_tangents = to_flat(args[1])
+        args_unwrapped = (args_unwrapped_primals, args_unwrapped_tangents)
+        wrap_info = (wrap_info_primals, wrap_info_tangents)
+    else:
+        args_unwrapped, wrap_info = to_flat(args)
+        args_unwrapped_primals, wrap_info_primals = args_unwrapped, wrap_info
+
+    if not any_subclass:
+        # No-op case
+        fx_g = trace_fn(flat_fn, args, meta=meta, aot_config=aot_config, trace_joint=trace_joint)
+        return fx_g, None
+
+
+    subclass_meta = TensorToSubclassInfo(
+        wrap_info=wrap_info,
+        maybe_joint_inputs=args_unwrapped,
+        # set later
+        meta=None,
+        fw_output_infos=None,
+        bw_output_infos=None,
+        fw_output_types=None,
+    )
+
+    # This is a function that takes in plain tensors, and figures out how to (potentially)
+    # wrap them into the tensor subclasses that our module expects.
+    # Order goes:
+    # def f(args_flat)
+    #     args_subclass = wrap(args_flat)
+    #     out_subclass = user_f(args_subclass)
+    #     out_flat = unwrap(out_subclass)
+    #     return out_flat
+    def get_wrapped_fn(fn, *, wrap_info_, store_output_meta: bool):
+        nonlocal subclass_meta
+
+        def wrapped_flat_fn(*args_unwrapped):
+            # In the joint case, wrap_info and args_unwrapped will not be given to us in a flattened manner;
+            # We'll get them both as (primals_list, tangents_list)
+            # This is annoying; I need to manually store some extra info for the primals and tangents separately,
+            # so I can't blindly use pytrees
+            # As a result, I'm effectively manually handling the calling conventions for the joint vs inference case.
+            if isinstance(args_unwrapped[0], list):
+                # joint case
+                assert len(args_unwrapped) == 2
+                assert isinstance(args_unwrapped[1], list)
+                assert len(wrap_info_) == 2
+                assert isinstance(wrap_info[0], list) and isinstance(wrap_info[1], list)
+                args_unwrapped_flat = args_unwrapped[0] + args_unwrapped[1]
+                is_joint = True
+            else:
+                args_unwrapped_flat = args_unwrapped
+                is_joint = False
+
+            args_wrapped = []
+            for infos in wrap_info_ if is_joint else [wrap_info_]:
+                curr_args_wrapped = []
+                for info in infos:
+                    if isinstance(info, int):
+                        curr_args_wrapped.append(args_unwrapped_flat[info])
+                        continue
+                    assert isinstance(info, SubclassCreationMeta)
```

Comment:
It smells like you have multiple copies of this logic lying around (grep for isinstance(info, SubclassCreationMeta)